File: leetcode/chain/143.py
# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def reverse(self, head: ListNode) -> None:
        ret = None
        while head:
            tmp = head.next
            head.next = ret
            ret = head
            head = tmp
        return ret  #返回自身

    def reorderList(self, head: ListNode) -> None:
        l = head

        slow, fast = head, head.next  #右半部分比左半部分更短，快慢指针求中间位置
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        m = slow.next
        slow.next = None

        m = self.reverse(m)
        while l and m:
            l_tmp = l.next
            m_tmp = m.next

            l.next = m
            m.next = l_tmp

            l = l_tmp
            m = m_tmp

File: leetcode/chain/143.cpp
#include <iostream>

#include "list_node.h"

class Solution {
 public:
  ListNode* reverse_list(ListNode* node) {
    ListNode* last = nullptr;
    ListNode* next = nullptr;
    while (node != nullptr) {
      next = node->next;
      node->next = last;

      last = node;
      node = next;
    }
    return last;
  }

  ListNode* reorder_list(ListNode* node) {
    ListNode *ret = node;
    // 快慢指针求中值
    ListNode *slow = node, *fast = node;
    while (fast && fast->next) {
      fast = fast->next->next;
      slow = slow->next;
    }
    ListNode *right = slow->next;
    slow->next = nullptr;
    // 右半边reverse
    ListNode* second = reverse_list(right);
    // 拼接
    while (node && second) {
      ListNode *node_next = node->next;
      ListNode *second_next = second->next;

      node->next = second;
      second->next = node_next;

      node = node_next;
      second = second_next;
    }

    return ret;
  }
};

int main() {
  ListNode* node = new ListNode(3);
  node->append(new ListNode(4));
  node->append(new ListNode(5));
  node->append(new ListNode(2));
  node->append(new ListNode(4));
  node->append(new ListNode(6));
  node->append(new ListNode(7));
  node->print();

  Solution solution = Solution();
  ListNode* ret = solution.reorder_list(node);
  ret->print();
}